Use bufio.Reader instead of bufio.Scanner to support messages with long lines

horejsek opened the pull request:
Message with very long lines (longer than `bufio.MaxScanTokenSize` which is `64 * 1024`) cannot be parsed using `bufio.Scanner`, see the [doc](https://golang.org/pkg/bufio/#Scanner):

> Programs that need more control over error handling or large tokens, or must run sequential scans on a reader, should use bufio.Reader instead.

To support also long lines, `bufio.Reader` should be used.

horejsek committed:
Use bufio.Reader instead bufio.Scanner to support messages with long lines

emersion commented:
Per [RFC 5322 section 2.1.1](https://tools.ietf.org/html/rfc5322#section-2.1.1), messages MUST have line lengths < 998 bytes.

Is it common to have non-conformant mbox files in practice? Do you know what software generates them?

horejsek commented:
Yep, we face this issue, Google generates such files...

Even though this is not fine by RFC, software should be able to cope with it if possible. It could be bug on their side but user experience that nothing can be read after such long line is not nice.

emersion commented:
Yes, I agree.

However, if we add support for long lines, I don't think we should do unbounded memory allocations. `bufio.Scanner` is a nice safeguard, `bufio.Reader.ReadBytes` doesn't have a limit. This means a single-line base64-encoded attachment may result in tens of megabytes allocations.

The mbox reader is only interested in:

- Converting newlines to CRLF
- Looking for lines beginning with "From "

So we don't actually need to read whole lines.

emersion commented on reader_test.go:
```diff
@@ -423,6 +423,30 @@ Content-Transfer-Encoding: 8bit
+Date: Thu, 01 Jan 2015 00:00:01 +0100
+Subject: Test
+
+This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a very long line. This is a ...
```
Hmm, please use `strings.Repeat` here instead.

horejsek replied:
Done, good idea. 👍

horejsek commented:
Do you think changing the whole algorithm of Read method (and making it less readable) worth to do for few unique mbox files?

emersion commented:
I think unbounded memory allocations in libraries are a security issue. Think of a server allowing users to import mbox files for instance.

horejsek commented:
Good point with the server.

In that case, we will use our forked version and we will follow up later as this is not priority for us at the moment.

emersion commented:
Sure, no problem. I'll see if I can give it a stab sometime.

horejsek commented:
Ah, my original fix was working only with files with line ending `\n` and not `\r\n`. 🤦  I fixed it to use `ReadLine` and changed the algo to be also memory safe. In the end it's not that much complicated how I was thinking it will be.

emersion reviewed (approved):
Look good to me, thanks!

Sorry for taking a long time to review this.
